jardin/instrumentation/notifier.py

```diff
@@ -0,0 +1,46 @@
+import os
+import uuid
+import queue
+import atexit
+import threading
+
+from jardin import config as config
+from jardin.instrumentation.base_subscriber import BaseSubscriber
+from jardin.instrumentation.event import Event
+
+class Notifer:
+  _subscribers = {}
+
+  @classmethod
+  def is_enabled(self):
+      return os.environ.get("JARDIN_INSTRUMENTATION_ENABLED", "false") == "true"
+
+  @classmethod
+  def subscribe(self, handler: BaseSubscriber) -> uuid.UUID:
+      if not self.is_enabled():
+          return
+      subscriber_id = uuid.uuid4()
+      if not isinstance(handler, BaseSubscriber):
+          raise TypeError("Handler must be subclass of BaseSubscriber")
+      self._subscribers[subscriber_id] = handler
+      return subscriber_id
+
+  @classmethod
+  def unsubscribe(self, subscriber_id: uuid.UUID):
+      if not self.is_enabled():
+          return
+      del self._subscribers[subscriber_id]
+
+  @classmethod
+  def report_event(self, ev: Event):
+      subscribers = dict(self._subscribers)
+      for subscriber_id in dict(self._subscribers):
+          subscriber = subscribers[subscriber_id]
```

Comment:
This is confusing. It looks like you are making a shallow copy to avoid external mutation while looping over the subscribers?

I haven't yet reviewed this PR's thread-safety, but I would consider using a lock.

Reply:
That is correct. 
I wanted to avoid adding locks since the race condition is fairly benign (we miss notifying a subscriber that has just been added)

Will add the lock just to avoid the confusion.